code/game/machinery/keypads.dm

```diff
@@ -0,0 +1,271 @@
+			password = ""
+		return
+	..()
+
+/obj/item/wallframe/keypad/attach(turf/on_wall) //I hate how I have to copy-paste code just to prevent the wallframe from being deleted before actually transferring variables.
+	var/pass = password
+	var/obj/machinery/keypad/O = ..()
+	O.password = pass
+
+//Machinery
+/obj/machinery/keypad
+	name = "keypad"
+	desc = "A remote control switch requiring a code input."
+	icon_state = "keypad"
+	var/password = ""
+	var/input = ""
+	var/output = "" //what to display in place of input
+	var/max_digits = 6
+	var/obj/item/device/assembly/device //Device pulsed on input success
+	var/obj/item/device/assembly/device2 //Device pulsed on bad password input
+	var/device_type = null
+	var/id = null
+	var/wired = 1
+
+/obj/machinery/keypad/New(loc, ndir = 0, built = 0)
+	..()
+	if(built)
+		dir = ndir
+		pixel_x = (dir & 3)? 0 : (dir == 4 ? -24 : 24)
+		pixel_y = (dir & 3)? (dir ==1 ? -24 : 24) : 0
+		panel_open = 1
+		wired = 0
+		stat |= MAINT
+		update_icon()
+
+	if(id && !built && !device && device_type)
+		device = new device_type(src)
+
+	if(id && istype(device, /obj/item/device/assembly/control))
+		var/obj/item/device/assembly/control/A = device
+		A.id = id
+
+/obj/machinery/keypad/update_icon()
+	overlays.Cut()
+	if(panel_open)
+		icon_state = "keypad-open"
+		if(wired)
+			overlays += "keypad-wired"
+		if(device)
+			overlays += "keypad-device"
+		if(device2)
+			overlays += "keypad-device2"
+	else
+		if(stat & (NOPOWER|BROKEN|MAINT))
+			icon_state = "[initial(icon_state)]-p"
+		else
+			icon_state = initial(icon_state)
+
+/obj/machinery/keypad/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/device/detective_scanner))
+		return
+
+	if(istype(W, /obj/item/weapon/screwdriver))
+		default_deconstruction_screwdriver(user, "keypad-open", "[initial(icon_state)]",W)
+		update_icon()
+		return
+
+	if(panel_open)
+		if(!wired && istype(W, /obj/item/stack/cable_coil))
+			var/obj/item/stack/cable_coil/C = W
+			if(C.get_amount() < 5)
+				user << "<span class='warning'>You need five lengths of cable for \the [src]!</span>"
+				return
+			user.visible_message("<span class='warning'>[user] adds cables to \the [src] frame.</span>", \
+								"<span class='notice'>You start adding cables to \the [src] frame...</span>")
+			playsound(loc, 'sound/items/Deconstruct.ogg', 50, 1)
+			if(do_after(user, 20, target = src))
+				if (C.amount >= 5)
+					C.use(5)
+					wired = 1
+					stat &= ~MAINT
+					user << "<span class='notice'>You add cables to \the [src] frame.</span>"
+					update_icon()
+		if((!device || !device2) && istype(W, /obj/item/device/assembly))
+			var/list/choices = list()
+			if(!device)
+				choices += "Code Input Success"
+			if(!device2)
+				choices += "Code Input Failure"
+			var/get_input = input("On what condition would this assembly be pulsed?","Device") as null|anything in choices
+			if(!in_range(src, user))
+				return
+			if(!user.canUseTopic(src))
```

Comment:
Nevermind, doing a canUseTopic twice is unneccesary since Attackby wouldn't be called anyway when they're on the ground or whatever